drag_and_drop_v2/drag_and_drop_v2.py

```diff
@@ -412,12 +414,224 @@ def default_background_image_url(self):
+
+        self._publish_item_dropped_event(item_attempt, is_correct)
+
+        return {}
+
+    def _validate_drop_item(self, item):
+        """
+        Validates `drop_item` parameters
+        """
+        zone = self._get_zone_by_uid(item['zone'])
+        if not zone:
+            raise JsonHandlerError(400, "Item zone data invalid.")
+
+    @staticmethod
+    def _make_state_from_attempt(attempt, correct):
+        """
+        Converts "attempt" data coming from browser into "state" entry stored in item_state
+        """
+        return {
+            'zone': attempt['zone'],
+            'correct': correct,
+            'x_percent': attempt['x_percent'],
+            'y_percent': attempt['y_percent'],
+        }
+
+    def _mark_complete_and_publish_grade(self):
+        """
+        Helper method to update `self.comnpleted` and submit grade event if appropriate conditions met.
+        """
+        if not self.completed or (self.mode == self.ASSESSMENT_MODE and not self.attempts_remain):
+            self.completed = self._is_finished() or not self.attempts_remain
+            grade = self._get_grade()
+            if grade > self.grade:
+                self.grade = grade
+                self._publish_grade()
+
+    def _publish_grade(self):
+        """
+        Publishes grade
+        """
+        try:
+            self.runtime.publish(self, 'grade', {
+                'value': self.grade,
+                'max_value': self.weight,
+            })
+        except NotImplementedError:
+            # Note, this publish method is unimplemented in Studio runtimes,
+            # so we have to figure that we're running in Studio for now
+            pass
+
+    def _publish_item_dropped_event(self, attempt, is_correct):
+        """
+        Publishes item dropped event.
+        """
+        item = self._get_item_definition(attempt['val'])
+        # attempt should already be validated here - not doing the check for existing zone again
+        zone = self._get_zone_by_uid(attempt['zone'])
+
+        self.runtime.publish(self, 'edx.drag_and_drop_v2.item.dropped', {
+            'item_id': item['id'],
+            'location': zone.get("title"),
+            'location_id': zone.get("uid"),
+            'is_correct': is_correct,
+        })
+
+    def _is_attempt_correct(self, attempt):
+        """
+        Check if the item was placed correctly.
+        """
+        correct_zones = self._get_item_zones(attempt['val'])
+        return attempt['zone'] in correct_zones
+
+    def _expand_static_url(self, url):
+        """
+        This is required to make URLs like '/static/dnd-test-image.png' work (note: that is the
+        only portable URL format for static files that works across export/import and reruns).
+        This method is unfortunately a bit hackish since XBlock does not provide a low-level API
+        for this.
+        """
+        if hasattr(self.runtime, 'replace_urls'):
```

Comment:
This method is pretty complicated. Hopefully there is test coverage for it (haven't gotten to the tests yet).